src/node_report.cc

```diff
@@ -0,0 +1,346 @@
+  #include <unistd.h>
+  #include <sys/resource.h>
+  #include <inttypes.h>
+#endif
+
+namespace node {
+
+using v8::HandleScope;
+using v8::HeapStatistics;
+using v8::HeapSpaceStatistics;
+using v8::Isolate;
+using v8::Local;
+using v8::Message;
+using v8::Object;
+using v8::String;
+using v8::StackFrame;
+using v8::StackTrace;
+using v8::V8;
+
+static void PrintStackFromMessage(FILE* fd, Isolate* isolate, Local<Message> message_handle);
+static void PrintStackFromStackTrace(FILE* fd, Isolate* isolate, DumpEvent event);
+static void PrintStackFrame(FILE* fd, Isolate* isolate, Local<StackFrame> frame, int index);
+static void PrintGCStatistics(FILE *fd, Isolate* isolate);
+static void PrintSystemInformation(FILE *fd, Isolate* isolate);
+static void WriteIntegerWithCommas(FILE *fd, size_t value);
+
+static int seq = 0; // sequence number for NodeReport filenames
+const char* v8_states[] = {"JS","GC","COMPILER","OTHER","EXTERNAL","IDLE"};
+
+#ifdef __POSIX__
+static struct {const char* description; int id;} rlimit_strings[] = {
+  {"core file size (blocks)       ", RLIMIT_CORE},
+  {"data seg size (kbytes)        ", RLIMIT_DATA},
+  {"file size (blocks)            ", RLIMIT_FSIZE},
+  {"max locked memory (bytes)     ", RLIMIT_MEMLOCK},
+  {"max memory size (kbytes)      ", RLIMIT_RSS},
+  {"open files                    ", RLIMIT_NOFILE},
+  {"stack size (bytes)            ", RLIMIT_STACK},
+  {"cpu time (seconds)            ", RLIMIT_CPU},
+  {"max user processes            ", RLIMIT_NPROC},
+  {"virtual memory (kbytes)       ", RLIMIT_AS}
+};
+#endif
+
+/*
+ * API to write a triage dump (NodeReport) to file
+ */
+void TriggerNodeReport(Isolate* isolate, DumpEvent event, const char *message, const char *location) {
+  TriggerNodeReport(isolate, event, message, location, Local<Message>());
+}
+
+/*
+ * API to write a triage dump (NodeReport) to file - with additional V8 Message supplied
+ */
+void TriggerNodeReport(Isolate* isolate, DumpEvent event, const char *message, const char *location, Local<Message> message_handle) {
+
+  // Construct the NodeReport filename, with timestamp, pid and sequence number
+  char filename[48] = "NodeReport";
+  seq++; // TODO: sequence number atomic?
+
+#ifdef _WIN32
+  SYSTEMTIME tm_struct;
+  GetLocalTime(&tm_struct);
+  DWORD pid = GetCurrentProcessId();
+
+  sprintf(&filename[10], ".%4d%02d%02d.%02d%02d%02d.%d.%03d.txt",
+          tm_struct.wYear,tm_struct.wMonth,tm_struct.wDay,tm_struct.wHour,tm_struct.wMinute,tm_struct.wSecond,pid,seq);
+#else // UNIX
+  struct timeval time_val;
+  struct tm *tm_struct;
+  gettimeofday(&time_val, NULL);
+
+  pid_t pid = getpid();
+
+  tm_struct = localtime(&time_val.tv_sec);
+  sprintf(&filename[10],".%4d%02d%02d.%02d%02d%02d.%d.%03d.txt",
+          tm_struct->tm_year+1900,tm_struct->tm_mon+1,tm_struct->tm_mday,tm_struct->tm_hour,tm_struct->tm_min,tm_struct->tm_sec,pid,seq);
+#endif
+
+  // Open the NodeReport file for writing
+  FILE *fd = fopen(filename, "w");
```

Comment:
Yes that's misleading, will change throughout. Also I should be checking for errors on the fopen().